src/modules/silentpayments/main_impl.h

```diff
@@ -0,0 +1,629 @@
+    (void)ser_ret;
+    memcpy(&public_data->data[1 + pubkeylen], input_hash_local, 32);
+    return ret;
+}
+
+static int secp256k1_silentpayments_recipient_public_data_load_pubkey(const secp256k1_context *ctx, secp256k1_pubkey *pubkey, const secp256k1_silentpayments_public_data *public_data) {
+    size_t pubkeylen = 65;
+    return secp256k1_ec_pubkey_parse(ctx, pubkey, &public_data->data[1], pubkeylen);
+}
+
+static void secp256k1_silentpayments_recipient_public_data_load_input_hash(unsigned char *input_hash, const secp256k1_silentpayments_public_data *public_data) {
+    size_t pubkeylen = 65;
+    memcpy(input_hash, &public_data->data[1 + pubkeylen], 32);
+}
+
+int secp256k1_silentpayments_recipient_public_data_serialize(const secp256k1_context *ctx, unsigned char *output33, const secp256k1_silentpayments_public_data *public_data) {
+    secp256k1_pubkey pubkey;
+    unsigned char input_hash[32];
+    size_t pubkeylen = 33;
+    int ret = 1;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(output33 != NULL);
+    ARG_CHECK(public_data != NULL);
+    /* Only allow public_data to be serialized if it has the hash and the summed public key
+     * This helps protect against accidentally serialiazing just a the summed public key A
+     */
+    ARG_CHECK(public_data->data[0] == 0);
+    ret &= secp256k1_silentpayments_recipient_public_data_load_pubkey(ctx, &pubkey, public_data);
+    secp256k1_silentpayments_recipient_public_data_load_input_hash(input_hash, public_data);
+    ret &= secp256k1_ec_pubkey_tweak_mul(ctx, &pubkey, input_hash);
+    secp256k1_ec_pubkey_serialize(ctx, output33, &pubkeylen, &pubkey, SECP256K1_EC_COMPRESSED);
+    return ret;
+}
+
+int secp256k1_silentpayments_recipient_public_data_parse(const secp256k1_context *ctx, secp256k1_silentpayments_public_data *public_data, const unsigned char *input33) {
+    size_t inputlen = 33;
+    size_t pubkeylen = 65;
+    secp256k1_pubkey pubkey;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(public_data != NULL);
+    ARG_CHECK(input33 != NULL);
+    /* Since an attacker can send us malicious data that looks like a serialized public key but is not, fail early */
+    if (!secp256k1_ec_pubkey_parse(ctx, &pubkey, input33, inputlen)) {
+        return 0;
+    }
+    public_data->data[0] = 1;
+    secp256k1_ec_pubkey_serialize(ctx, &public_data->data[1], &pubkeylen, &pubkey, SECP256K1_EC_UNCOMPRESSED);
+    memset(&public_data->data[1 + pubkeylen], 0, 32);
+    return 1;
+}
+
+int secp256k1_silentpayments_recipient_scan_outputs(
+    const secp256k1_context *ctx,
+    secp256k1_silentpayments_found_output **found_outputs, size_t *n_found_outputs,
+    const secp256k1_xonly_pubkey * const *tx_outputs, size_t n_tx_outputs,
+    const unsigned char *recipient_scan_key,
+    const secp256k1_silentpayments_public_data *public_data,
+    const secp256k1_pubkey *recipient_spend_pubkey,
+    const secp256k1_silentpayments_label_lookup label_lookup,
+    const void *label_context
```

Comment:
Could label_context (here and in the callback type) be made mutable? Is there a particular reason it is `const`?

Reply:
Does this actually disallow (make undefined behavior) mutating the context from the callback?

Reply:
What would be the reason for making the pointer to label context mutable here? Making it const is meant to clearly communicate to the caller that this is meant to be a reference to some external data store. For your second question, I don't think making the pointer const prevents the lookup function itself from mutating the context, but I don't think there is a way we can prevent that.

Reply:
If I understand correctly (after searching online about this), the C compiler will assume that no mutation happens to the pointed to data using the pointer `label_context`, because it is declared `const`.
In Rust however, you can't prevent users from mutating captured variables of a closure. (Rust's shared references can be mutated through, still they are called immutable references.)

A fix on the Rust side would be to add a level of indirection, instead of passing a pointer to the context directly, a pointer to a pointer is passed. The pointed to data of the `const void *` is never mutated (that would of type `context_t *` then, in C terms, where `context_t` is a struct defined by the user).
A fix on the C side would be to remove the `const` keyword.

I am fine with both, but the extra level of indirection will be slightly less performant, but modern CPU caches can make this (almost) unobservable.

Reply:
Thanks for the clarification!

> If the object itself is non-`const`, but you hold a `const` pointer to it, then it is perfectly legal even to cast the `const`ness of the pointer away and mutate it through that casted pointer.

Especially this, this is new for me, thanks!